Do not shutdown bitcoind if ActivateBestChain() fails on startup

ptschip opened the pull request:
Log a message instead and continue to startup. This allows the node
to start and the operator to run commands such as getchaintips or
reconsidermostworkchain so they can then try to get the node startup
to complete.

Furthermore, we run reconsidermostworkchain() both before and after
ActivateBestChain(). We do it before in the case where a previous
shutdown left the chain in an invalid state, and we do it after, in
the case where an operator is switchin from an ABC/BCHN client to
a BU client (In this situation an ABC client could have parked a few
blocks which would make the BU node hang on startup).

ptschip committed:
Do not shutdown bitcoind if ActivateBestChain() fails on startup

Log a message instead and continue to startup. This allows the node
to start and the operator to run commands such as getchaintips or
reconsidermostworkchain so they can then try to get the node startup
to complete.

Furthermore, we run reconsidermostworkchain() both before and after
ActivateBestChain(). We do it before in the case where a previous
shutdown left the chain in an invalid state, and we do it after, in
the case where an operator is switchin from an ABC/BCHN client to
a BU client (In this situation an ABC client could have parked a few
blocks which would make the BU node hang on startup).

ptschip committed:
During startup put a try/catch block around reconsidermostworkchain.

ptschip committed:
Check each fork for the correct one if we have a tie in nChainWork

ptschip committed:
Add test to blockchain.py

sickpig commented on src/init.cpp:
```diff
@@ -427,6 +427,30 @@ void CleanupBlockRevFiles()
+        std::set<CBlockIndex *, CompareBlocksByHeight> setTips;
+        {
+            LOCK(cs_main);
+            setTips = GetChainTips();
```
this seems to be unused in this function.

ptschip replied:
oops, was left over from somewhere...thanks.

sickpig commented on src/rpc/blockchain.cpp:
```diff
@@ -1620,30 +1620,63 @@ UniValue reconsidermostworkchain(const UniValue &params, bool fHelp)
+        if (pTipToVerify->nChainWork == chainActive.Tip()->nChainWork)
+        {
+            LOGA("Nothing to do. Already on the correct chain.");
+            throw runtime_error("Nothing to do. Already on the correct chain.");
```
wouldn't this mean that when we hit this behavior while the function is run on the startup phase (`init.cpp`) we are going to exit `bitcoind`/`bitcoin-qt?`

ptschip replied:
No it won't exit because when we call it in init.cpp it's in a try/catch block.

sickpig replied:
you are right indeed. thanks!

ptschip committed:
Remove unused code

gandrewstone commented on src/rpc/blockchain.cpp:
```diff
@@ -1620,30 +1620,63 @@ UniValue reconsidermostworkchain(const UniValue &params, bool fHelp)
+            // Set the rollbackchain override flag if there was one provided.
+            obj.push_back(params[0]);
+        }
+        rollbackchain(obj, false);
```
RPC call functions should only be called from the RPC interface.  Please make a new function that takes normal arguments, call it here, and have the rollbackchain implementation also call your new function.  (I think there's more than just rollbackchain that needs this to be done).

ptschip replied:
done

ptschip committed:
Create separate functions for rollbackchain and reconsidermostworkchain

which can be called from inside and outside the rpc calls. This gives
us a cleaner and less error prone way of making the calls outside of
of the rpc interface.

Errors are returned as strings which can then be used to either
log a message in the debug log or, as in the case of the rpc interface,
throw an informative exception which can be returned to the end user.

t
